packages/router/contracts/adapters/PendleAdapter.sol

```diff
@@ -0,0 +1,427 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity 0.8.19;
+
+import '@pendle/core-v2/contracts/router/base/MarketApproxLib.sol';
+import '@pendle/core-v2/contracts/interfaces/IPMarket.sol';
+import '@pendle/core-v2/contracts/core/StandardizedYield/PYIndex.sol';
+
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+import '@uniswap/v3-periphery/contracts/libraries/TransferHelper.sol';
+
+import '../interfaces/IMarginlyRouter.sol';
+
+contract PendleAdapter {
+  using PYIndexLib for IPYieldToken;
+
+  event NewPair(address indexed token0, address indexed token1, address pendleMarket, address uniswapV3LikePool);
+
+  error UnknownPool();
+  error InsufficientAmount();
+  error TooMuchRequested();
+  error Forbidden();
+
+  struct PendleMarketData {
+    IPMarket market;
+    IStandardizedYield sy;
+    IPPrincipalToken pt;
+    IPYieldToken yt;
+    address ib;
+    address uniswapV3;
+  }
+
+  struct PoolData {
+    address pendleMarket;
+    address uniswapV3LikePool;
+    address ib;
+  }
+
+  struct PoolInput {
+    PoolData poolData;
+    address tokenA;
+    address tokenB;
+  }
+
+  struct CallbackData {
+    address tokenIn;
+    address tokenOut;
+    address syToken;
+    address initiator;
+    uint256 maxAmountIn;
+    bytes data;
+  }
+
+  uint256 private constant EPSILON = 1e15;
+  uint256 private constant ONE_PLUS_SLIPPAGE = 105;
+  uint256 private constant ONE_MINUS_SLIPPAGE = 95;
+  uint256 private constant MAX_ITERATIONS = 10;
```

Comment:
Consider using another types (lower than uint256) for storing these constants. would it be better for contract size?

Reply:
I checked, type conversion increases the size more than using uint256

Reply:
> I checked, type conversion increases the size more than using uint256

Not really true: I use `private` constants, type conversion doesn't matter in that case and is definitely resolved at compilation, as the result, nos significant shift in contract size should occur.

That's said, in case of `public` constants type conversion has more drastic impact on contract size, though not that big of a deal. I'm not 100% sure of the way they're stored, but I'd guess it acts more or less the same way as state variable